src/request/sign-transaction/SignTransactionApi.js

```diff
@@ -4,26 +4,71 @@
 /* global Nimiq */
 /* global AmountTooSmallError, NetworkMissmatchError */
 /* global SignTransactionWithPassphrase, SignTransactionWithPin */
+/* global TransactionType */
+/* global AddressUtils */
+/* global Utf8Tools */
 
 class SignTransactionApi extends PopupApi { // eslint-disable-line no-unused-vars
     /**
      * @param {TransactionRequest} txRequest
      */
     async onRequest(txRequest) {
-        if (Nimiq.Policy.coinsToSatoshis(txRequest.value) < 1) {
+        // Validation
+        if (!txRequest) throw new Error('Invalid request');
+
+        // These checks throw an exception on failure
+        AddressUtils.isUserFriendlyAddress(txRequest.sender);
+        AddressUtils.isUserFriendlyAddress(txRequest.recipient);
+        AddressUtils.isUserFriendlyAddress(txRequest.signer);
+        if (typeof txRequest.extraData !== 'string') {
+            throw new Error('Transaction extraData must be a string');
+        }
+
+        // Normalization
+        txRequest.value = Nimiq.Policy.coinsToSatoshis(txRequest.value);
+        txRequest.fee = Nimiq.Policy.coinsToSatoshis(txRequest.fee);
+        txRequest.sender = AddressUtils.formatAddress(txRequest.sender);
+        txRequest.recipient = AddressUtils.formatAddress(txRequest.recipient);
+        txRequest.signer = AddressUtils.formatAddress(txRequest.signer);
+        txRequest.extraData = Utf8Tools.utf8ByteArrayToString(
+            Utf8Tools.stringToUtf8ByteArray(txRequest.extraData || ''),
+        );
+
+        switch (txRequest.type) {
+        case TransactionType.BASIC:
+            if (txRequest.sender !== txRequest.signer) {
+                throw new Error('Sender must be signer for basic transactions');
+            }
+
+            if (txRequest.extraData) {
+                throw new Error('Cannot add extraData to basic transaction');
+            }
+
+            break;
+        case TransactionType.EXTENDED:
+            // Validate extended transactions
+
+            break;
+        default:
+            throw new Error('Invalid transaction type');
+        }
+
+        if (txRequest.value < 1) {
             throw new AmountTooSmallError();
         }
 
         if (txRequest.network !== Nimiq.GenesisConfig.NETWORK_NAME) {
             throw new NetworkMissmatchError(txRequest.network, Nimiq.GenesisConfig.NETWORK_NAME);
         }
+        // End validation
 
-        txRequest.value = Nimiq.Policy.coinsToSatoshis(txRequest.value);
-        txRequest.fee = Nimiq.Policy.coinsToSatoshis(txRequest.fee);
-
-        // get key from keystore
-        const keyStore = KeyStore.instance;
-        const keyType = await keyStore.getType(txRequest.sender);
+        // Get signer key type
+        let keyType;
+        if (txRequest.mockKeyType) keyType = txRequest.mockKeyType;
```

Comment:
what about this mockKeyType? What is it for?

Reply:
For testing from the KeyguardClient, as long as we do not have stored keys.

Reply:
Is there no way to put this into testing code? It pollutes the general code if it's for testing purpose only.